File: rails/app/myapp/app/models/product.rb
# frozen_string_literal: true

# Product Model
class Product
  include Mongoid::Document

  field :sku, type: String
  field :name, type: String
  field :description, type: String
  field :amount, type: Integer
  field :price, type: Float

  validates :sku, presence: true
  validates :name, presence: true
  validates :description, presence: true
  validates :amount, presence: true
  validates :price, presence: true

  def as_json(*_args)
    h = super()
    h['id'] = h.delete('_id').to_s
    h.as_json(only: %w[id sku name description amount price])
  end
end

File: phoenix/app/myapp/lib/myapp/management/product.ex
defmodule Myapp.Management.Product do
  @moduledoc """
  Product model
  """

  use Ecto.Schema

  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}

  schema "products" do
    field(:amount, :integer)
    field(:description, :string)
    field(:name, :string)
    field(:price, :float)
    field(:sku, :string)
    field(:barcode, :string)

    timestamps()
  end

  @doc false
  def changeset(product, attrs) do
    product
    |> cast(attrs, [:sku, :amount, :description, :name, :price, :barcode])
    |> validate_required([:sku, :amount, :description, :name, :price, :barcode])
    |> validate_format(:sku, ~r/^([a-zA-Z0-9]|\-)+$/,
      message: "can only contain alphanumerics and hifen"
    )
    |> validate_number(:price, greater_than: 0)
    |> validate_length(:barcode, min: 8, max: 13)
  end

  def get_attrs(product) do
    %{
      amount: product.amount,
      barcode: product.barcode,
      description: product.description,
      id: product.id,
      name: product.name,
      price: product.price,
      sku: product.sku
    }
  end
end
